File: android/app/src/main/java/com/noop/protocol/Whoop5RR.kt
package com.noop.protocol

/** WHOOP 5 type-40/v18 words are 1/1024-second ticks. Swift Whoop5RR; WHOOP 4 stays unchanged. */
object Whoop5RR {
    fun milliseconds(ticks: Int): Int {
        require(ticks in 0..65535)
        return (ticks * 1000 + 512) / 1024
    }

    fun usesCanonicalSource(model: String?, brand: String?, hasTaggedIntervals: Boolean): Boolean {
        if (!brand.isNullOrEmpty() && !brand.equals("WHOOP", ignoreCase = true)) return false
        return when (DeviceFamily.confirmedRegistryFamily(model, brand)) {
            DeviceFamily.WHOOP4 -> false
            DeviceFamily.WHOOP5 -> true
            null -> hasTaggedIntervals
        }
    }

}

File: Packages/WhoopProtocol/Sources/WhoopProtocol/Whoop5RR.swift
import Foundation

/// WHOOP 5 type-40 and v18 interval words use the same 1/1024-second ticks as BLE 0x2A37.
/// Verified by matching complete native/standard beat arrays on firmware 50.41.1.0.
/// Keep this separate from WHOOP 4 decoding, whose existing millisecond contract is unchanged.
public enum Whoop5RR {
    public static func milliseconds(ticks: UInt16) -> Int {
        (Int(ticks) * 1000 + 512) / 1024
    }

    /// Labelled wire observations resolve an unknown registry entry, but cannot override another family.
    public static func usesCanonicalSource(model: String?, brand: String?, hasTaggedIntervals: Bool) -> Bool {
        if let brand, !brand.isEmpty, brand.caseInsensitiveCompare("WHOOP") != .orderedSame { return false }
        switch DeviceFamily.confirmedRegistryFamily(model: model, brand: brand) {
        case .whoop4: return false
        case .whoop5: return true
        case nil: return hasTaggedIntervals
        }
    }

}
